From f16b525c94d12bced2f43e6799225ca828360c78 Mon Sep 17 00:00:00 2001
From: hyphop <art@khadas.com>
Date: Sat, 25 Apr 2020 23:02:57 +0900
Subject: [PATCH] khadas_logo

+ logo + preboot
+ PREBOOT => load_logo + usb_start + kbi_init + sleep 1 sec
+ add missed $loadaddr value
+ bmp logo file can be compressed by gzip
+ load LOGO seq
	 -> SPI
	 -> MMC|SD /splash.bmp
	 -> MMC|SD /usr/share/fenix/logo/logo.bmp
	 -> embeded logo $fdtcontroladdr_end

---
 configs/khadas-edge-captain-rk3399_defconfig |  2 ++
 configs/khadas-edge-rk3399_defconfig         |  2 ++
 configs/khadas-edge-v-rk3399_defconfig       |  2 ++
 configs/khadas-vim2_defconfig                |  2 ++
 configs/khadas-vim3_defconfig                |  2 ++
 configs/khadas-vim3l_defconfig               |  2 ++
 configs/khadas-vim_defconfig                 |  2 ++
 include/configs/meson64.h                    | 11 +++++++++++
 8 files changed, 25 insertions(+)

diff --git a/configs/khadas-edge-captain-rk3399_defconfig b/configs/khadas-edge-captain-rk3399_defconfig
index 64b23492..845bfa69 100644
--- a/configs/khadas-edge-captain-rk3399_defconfig
+++ b/configs/khadas-edge-captain-rk3399_defconfig
@@ -19,6 +19,7 @@ 
 CONFIG_SPL_STACK_R_MALLOC_SIMPLE_LEN=0x10000
 CONFIG_TPL=y
 CONFIG_SYS_PROMPT="kedge# "
+CONFIG_USE_PREBOOT=y
 CONFIG_CMD_BOOTZ=y
 CONFIG_CMD_GPT=y
 CONFIG_CMD_MMC=y
diff --git a/configs/khadas-edge-rk3399_defconfig b/configs/khadas-edge-rk3399_defconfig
index b323a653..8bea7365 100644
--- a/configs/khadas-edge-rk3399_defconfig
+++ b/configs/khadas-edge-rk3399_defconfig
@@ -19,6 +19,7 @@ CONFIG_SPL_STACK_R=y
 CONFIG_SPL_STACK_R_MALLOC_SIMPLE_LEN=0x10000
 CONFIG_TPL=y
 CONFIG_SYS_PROMPT="kedge# "
+CONFIG_USE_PREBOOT=y
 CONFIG_CMD_BOOTZ=y
 CONFIG_CMD_GPT=y
 CONFIG_CMD_MMC=y
diff --git a/configs/khadas-edge-v-rk3399_defconfig b/configs/khadas-edge-v-rk3399_defconfig
index 652a15b9..931163a1 100644
--- a/configs/khadas-edge-v-rk3399_defconfig
+++ b/configs/khadas-edge-v-rk3399_defconfig
@@ -19,6 +19,7 @@ CONFIG_SPL_STACK_R=y
 CONFIG_SPL_STACK_R_MALLOC_SIMPLE_LEN=0x10000
 CONFIG_TPL=y
 CONFIG_SYS_PROMPT="kedge# "
+CONFIG_USE_PREBOOT=y
 CONFIG_CMD_BOOTZ=y
 CONFIG_CMD_GPT=y
 CONFIG_CMD_MMC=y
diff --git a/configs/khadas-vim2_defconfig b/configs/khadas-vim2_defconfig
index 30ce9d55..fe02e793 100644
--- a/configs/khadas-vim2_defconfig
+++ b/configs/khadas-vim2_defconfig
@@ -13,6 +13,7 @@ CONFIG_NR_DRAM_BANKS=1
 CONFIG_DEBUG_UART_BASE=0xc81004c0
 CONFIG_DEBUG_UART_CLOCK=24000000
 CONFIG_IDENT_STRING=" khadas-vim2"
+CONFIG_USE_PREBOOT=y
 CONFIG_DEBUG_UART=y
 CONFIG_OF_BOARD_SETUP=y
 CONFIG_MISC_INIT_R=y
diff --git a/configs/khadas-vim3_defconfig b/configs/khadas-vim3_defconfig
index 1bf1aa77..b0d55d67 100644
--- a/configs/khadas-vim3_defconfig
+++ b/configs/khadas-vim3_defconfig
@@ -21,6 +21,7 @@ CONFIG_NR_DRAM_BANKS=1
 CONFIG_DEBUG_UART_BASE=0xff803000
 CONFIG_DEBUG_UART_CLOCK=24000000
 CONFIG_IDENT_STRING=" khadas-vim3"
+CONFIG_USE_PREBOOT=y
 CONFIG_DEBUG_UART=y
 CONFIG_OF_BOARD_SETUP=y
 CONFIG_MISC_INIT_R=y
diff --git a/configs/khadas-vim3l_defconfig b/configs/khadas-vim3l_defconfig
index aab20da8..5f64b480 100644
--- a/configs/khadas-vim3l_defconfig
+++ b/configs/khadas-vim3l_defconfig
@@ -21,6 +21,7 @@ CONFIG_NR_DRAM_BANKS=1
 CONFIG_DEBUG_UART_BASE=0xff803000
 CONFIG_DEBUG_UART_CLOCK=24000000
 CONFIG_IDENT_STRING=" khadas-vim3l"
+CONFIG_USE_PREBOOT=y
 CONFIG_DEBUG_UART=y
 CONFIG_OF_BOARD_SETUP=y
 CONFIG_MISC_INIT_R=y
diff --git a/configs/khadas-vim_defconfig b/configs/khadas-vim_defconfig
index 93bab946..6e89ae54 100644
--- a/configs/khadas-vim_defconfig
+++ b/configs/khadas-vim_defconfig
@@ -13,6 +13,7 @@ CONFIG_NR_DRAM_BANKS=1
 CONFIG_DEBUG_UART_BASE=0xc81004c0
 CONFIG_DEBUG_UART_CLOCK=24000000
 CONFIG_IDENT_STRING=" khadas-vim"
+CONFIG_USE_PREBOOT=y
 CONFIG_DEBUG_UART=y
 CONFIG_OF_BOARD_SETUP=y
 CONFIG_MISC_INIT_R=y
diff --git a/include/configs/meson64.h b/include/configs/meson64.h
index 87ffc7c9..13988b87 100644
--- a/include/configs/meson64.h
+++ b/include/configs/meson64.h
@@ -75,12 +75,33 @@
 	func(DHCP, dhcp, na)
 #endif
 
+#define CONFIG_SYS_VIDEO_LOGO_MAX_SIZE 8192000
+#define CONFIG_VIDEO_BMP_GZIP 1
+
+#define CONSOLE_FONT_COLOR 14
+
+#define PREBOOT_LOAD_LOGO \
+	"ll=0; test $boot_source = spi && sf probe && sf read $loadaddr 0x170000 0x10000 && ll=1; " \
+	"test $ll = 0 && ll=1 && " \
+	"load mmc 1 $loadaddr splash.bmp || " \
+	"load mmc 1:2 $loadaddr /usr/share/fenix/logo/logo.bmp || " \
+	"load mmc 2 $loadaddr splash.bmp || " \
+	"load mmc 2:2 $loadaddr /usr/share/fenix/logo/logo.bmp || " \
+	"ll=0; " \
+	"test $ll = 1 && bmp display $loadaddr m m || ll=0; " \
+	"test $ll = 0 && bmp display $fdtcontroladdr_end m m; "
+
+#define PREBOOT_CMD "run load_logo; usb start; kbi init; sleep 1;"
+
 #ifndef CONFIG_EXTRA_ENV_SETTINGS
 #define CONFIG_EXTRA_ENV_SETTINGS \
+       "load_logo=" PREBOOT_LOAD_LOGO "\0" \
+       "preboot=" PREBOOT_CMD "\0" \
 	"stdin=" STDIN_CFG "\0" \
 	"stdout=" STDOUT_CFG "\0" \
 	"stderr=" STDOUT_CFG "\0" \
 	"fdt_addr_r=0x08008000\0" \
+	"loadaddr=0x01000000\0" \
 	"scriptaddr=0x08000000\0" \
 	"kernel_addr_r=0x08080000\0" \
 	"pxefile_addr_r=0x01080000\0" \
-- 
2.17.1

